### ISSUE TITLE
ButterCollection cannot be correctly marshal loaded because of dynamic method generation

### ISSUE THREAD

**Issue by bbugh** [ButterCMS/buttercms-ruby#13]
We're using Rails cache to store our Butter API call results. The dynamically generated methods on `ButterResource` are not restored when restored from cache, leading to unreliable errors like `NameError: undefined method `seo_title' for #<ButterCMS::Post:0x00007fec4faa2cb8>`. This happens any time the server restarts.

Because of the way ButterResource is dynamically creating methods, the methods are not restored when a `ButterCollection` is restored from cache. Ruby only calls initialize on the main object when restoring via `Marshal.load`. 

- When restoring a `ButterResource`, the `initialize` method is called. 
- When restoring a `ButterCollection`, the `initialize` method of the `ButterCollection` is called, but not the internal `ButterResource`s.

There's a couple different ways to fix this, which is why I am making an issue instead of a PR:

1. Define the `marshal_dump` and `marshal_load` methods on `ButterResource` to dynamically add the methods in `marshal_load`. I think this is the only solution that preserves the current behavior
2. Remove the dynamically defined methods and instead use `method_missing` to pass through the unknown calls to the `@data` `OpenStruct`. This could break existing code if people are checking for `respond_to?`

The marshal code would look something like this:

```ruby
    # unfortunately, Ruby expects you to define both methods for this to work, 
    # so the dump has to be manually specified as well
    def marshal_dump
      { json: @json, data: @data, meta: @meta }
    end

    def marshal_load(dump)
      @json = dump[:json]
      @data = dump[:data]
      @meta = dump[:meta]

      return unless @data.respond_to?(:each_pair)

      @data.each_pair do |key, value|
        instance_variable_set("@#{key}", @data.send(key))
        self.class.send(:attr_reader, key)
      end
    end
```

I expect that the first option is the best for preserving backwards compatibility, what do you think?

**Comment by jakelumetta** [comment]
Thakns @bbugh - @rposborne @kevinjcoleman; not sure if either of you are utilizing the cache capability? Thoughts on this issue / approach?

**Comment by jakelumetta** [comment]
@bbugh After reviewing, option 1 seems like the best approach. Backwards compatibility is important to maintain if possible.  Thanks!

**Comment by kevinjcoleman** [comment]
@jakelumetta @bbugh yeah I'm not utilizing the caching capabilities, but option 1 also seems like the best approach to me!

**Comment by bbugh** [comment]
Thank you both for your time, I will submit the PR today.

**Comment by bbugh** [comment]
Addressed in #14
